crates/typst-pdf/src/page.rs

```diff
@@ -711,6 +724,129 @@ fn write_group(ctx: &mut PageContext, pos: Point, group: &GroupItem) {
 
 /// Encode a text run into the content stream.
 fn write_text(ctx: &mut PageContext, pos: Point, text: &TextItem) {
+    // If the text run contains either only emojis or normal text
+    // we can render it directly
+    let sbix = match text.font.ttf().tables().sbix {
+        None => {
+            write_normal_text(ctx, pos, text);
+            return;
+        }
+        Some(sbix) => sbix,
+    };
+
+    let strike = sbix.best_strike(text.size.to_f32() as u16).unwrap();
+    let is_emoji = |g: &Glyph| strike.get(GlyphId(g.id)).is_some();
+    let emoji_count = text.glyphs.iter().filter(|g| is_emoji(*g)).count();
+
+    if emoji_count == text.glyphs.len() {
+        write_emojis(ctx, pos, text);
+    } else if emoji_count == 0 {
+        write_text(ctx, pos, text)
+    } else {
+        // Otherwise we need to split it in smaller text runs
+        let mut offset = 0;
+        let mut position_in_run = Em::zero();
+        while offset < text.glyphs.len() {
+            // Start a new text run where the last one ended
+            let start = offset;
+            // Determine if this is an emoji-only or a text-only run
+            let in_emoji_group = is_emoji(&text.glyphs[start]);
+            // Determine the index of the last glyph of the run
+            let end = start
+                + text.glyphs[start..]
+                    .iter()
+                    .position(|g| is_emoji(&g) != in_emoji_group)
+                    .unwrap_or(text.glyphs.len() - start);
+
+            // Build a sub text-run
+            let text_start = text.glyphs[start].range().start;
+            let text_end = text.glyphs[end - 1].range().end;
+
+            let new_text_item = TextItem {
```

Comment:
I think rather than creating a new text item, we could have some type like `TextItemView<'_>` or `SlicedTextItem<'_>`, which borrows the TextItem and knows the relevant subranges.